### PR TITLE
feat(embeddings): upgrade to Qwen3 model with performance optimizations

### PR DESCRIPTION
**Pull request by Ricardo1232**
### Model Upgrade (Primary Change)
- **Qwen3-Embedding-0.6B**: Upgraded from previous embedding model to Qwen3-Embedding-0.6B with 1024 dimensions
- **Improved quality**: Better embedding quality and semantic understanding
- **Model optimization**: Configured with proper PyTorch inference settings for optimal performance
- **Compatibility**: Updated transformers dependency to support the new model

### Performance Optimizations
- **Pandas to Polars migration**: Complete replacement of pandas DataFrames with Polars for improved CPU processing speed
- **PyTorch inference optimization**: Implementation of `model.eval()` and `torch.set_grad_enabled(False)` for memory efficiency
- **Memory management**: Wrapping inference calls with `torch.no_grad()` context manager for better memory usage
- **Combined benefit**: Model upgrade + optimizations deliver 22% performance improvement (5:36 minutes saved in test case)

### Database Schema Enhancement
- **Page number tracking**: Added `page_number` column to chunks table for better fragment manipulation
- **Improved indexing**: Enhanced database structure for future modifications of specific fragments
- **Better data access**: Migration from pandas `.iloc[0]['column']` to Polars `.row(0, named=True)['column']` syntax

### Code Quality Improvements
- **Text cleaning**: Added `clean_br()` function to remove `<br>` tags from text before embedding generation
- **Progress tracking**: Improved logging with tqdm for page-by-page progress visualization
- **Cleaner output**: Removed `show_progress_bar` from embedding generation for cleaner logs
- **Comment translation**: Updated Spanish comments to English for better code maintainability

### Technical Decisions
- **No batch processing in encoding**: PyTorch already utilizes all CPU cores internally for the `encode()` method, making batch processing unnecessary
- **Individual chunk processing**: Avoided batch database operations using `executemany()` due to memory escalation issues
- **Memory-first approach**: Chose simple PyTorch-level optimizations over complex batching mechanisms
- **Backward compatibility**: Maintained existing functionality while improving performance

#### Database Batch Processing Analysis
Initial implementation attempted batch processing with `executemany()` for database insertions (referenced in PR #33 ). However, this approach caused a memory snowball effect:
- **10 chunks**: ~3 GB RAM usage
- **30 chunks**: ~10 GB RAM usage  
- **50 chunks**: ~12 GB RAM usage

The exponential memory growth made the approach impractical for large documents, requiring complex memory management that would compromise code simplicity. Additionally, the memory pressure caused severe CPU performance degradation: during testing with a sample document, processing reached page 49 in approximately 6 minutes, but processing each subsequent chunk after page 50 took around 5 minutes per fragment, making the process practically unusable.

The current implementation prioritizes reliability and maintainability over theoretical performance gains, avoiding both memory and CPU bottlenecks.

### Dependencies Update
- **New dependencies**: Added `polars>=1.17.0`, `pyarrow>=18.0.0`, and `duckdb>=1.3.2`
- **Version updates**: Updated `transformers>=4.53.3` for Qwen model compatibility
- **Maintained compatibility**: Kept existing dependencies for other system components

### Test Conditions
- **Test case**: 166-page document (02012025-MAT)
- **Command**: `python extract_embeddings.py dof_md\2025\01\02012025-MAT`
- **Database state**: Fresh database for both tests to ensure identical conditions
- **Hardware**: Same system configuration for both runs

### Results
- **Previous performance**: 25:29 minutes with original script + Qwen model
- **New performance**: 19:53 minutes with optimized script + Qwen model
- **Performance improvement**: **22% reduction in processing time (5:36 minutes saved)**
- **Memory efficiency**: Reduced memory footprint through proper PyTorch inference configuration

The optimization delivers measurable performance gains while maintaining code simplicity and reliability.

### Migration Details
- **DataFrame creation**: Changed from `.df()` to `.pl()` for DuckDB query results
- **Data access patterns**: Updated from pandas syntax to Polars equivalents
- **Empty checks**: Replaced `.empty` with `.is_empty()` method
- **Column iteration**: Updated to use `.to_list()` for efficient column access

### User Experience Improvements
- **Better progress visibility**: Page-by-page processing feedback
- **Cleaner logs**: Removed redundant progress bars

### CODE REVIEW

File: extract_embeddings.py [hunk 1]
```diff
@@ -57,7 +58,13 @@
 # %%
 
 # Model configuration
-model = SentenceTransformer("nomic-ai/modernbert-embed-base", trust_remote_code=True)
+model = SentenceTransformer("Qwen/Qwen3-Embedding-0.6B", trust_remote_code=True, device="cpu")
```
Comment (jackbravo) [inline extract_embeddings.py:61]: I think the `device="cpu"` shouldn't be hardcoded. Some systems where we run this could have gpu or cuda.

File: extract_embeddings.py [hunk 2]
```diff
@@ -57,7 +58,13 @@
...
+
+# Set model to evaluation mode to save memory
+model.eval()
```
Comment (jackbravo) [inline extract_embeddings.py:64]: where did you learned this? Maybe paste the link in the comment?

File: extract_embeddings.py [hunk 3]
```diff
@@ -453,7 +465,12 @@ def _generate_chunk_embedding(header: str, chunk_text: str, file_path: str, chun
     text_for_embedding = f"{header}\n\n{chunk_text}{description_images}"
     
     try:
-        embedding = model.encode(f"search_document: {text_for_embedding}")
+        # Best practice: wrap inference calls in torch.no_grad()
+        with torch.no_grad():
+            embedding = model.encode(f"search_document: {text_for_embedding}", show_progress_bar=False)
```
Comment (jackbravo) [inline extract_embeddings.py:470]: Same comment as above about source for this.

File: extract_embeddings.py [hunk 4]
```diff
@@ -221,6 +229,10 @@ def build_header_dict(doc_title: str, page: str, open_headings: Dict[int, str])
     
     return "\n".join(header_lines)
 
+# --- Function to clean <br> tags ---
+def clean_br(text: str) -> str:
+    return text.replace('<br>', '')
+
```
Comment (jackbravo) [inline extract_embeddings.py:235]: Did something happen that we now have more of this?

**Review by jackbravo — APPROVED** [review]
LGTM with comments, please address them before merging.

**Comment by jackbravo** [comment]
Added #35 related to above.

File: extract_embeddings.py (hunk 1 above)
Comment (Ricardo1232) [reply extract_embeddings.py]: Got it, that modification is already in commit https://github.com/CodeandoGuadalajara/dof-rag/pull/34/commits/c88bc839e23459a1cf3b358fee5979f352750981

**Comment by Ricardo1232** [comment]
**Response regarding `eval` usage and PyTorch optimizations:**

With the upgrade to **Qwen3-Embedding-0.6B**, I introduced specific PyTorch optimizations to reduce memory usage and improve inference performance since it is a larger model than the previous one:

## 1. `model.eval()`
- Switches the model to evaluation mode, disabling training-specific layers such as *dropout* and *batch norm*.
- Reference: [PyTorch model.eval() vs torch.no_grad()](https://stackoverflow.com/questions/55627780/evaluating-pytorch-models-with-torch-no-grad-vs-model-eval)

## 2. `torch.set_grad_enabled(False)`
- Globally disables gradient computation for the entire session.
- Significantly reduces memory consumption by avoiding autograd tracking.
- Reference: [model.eval() with torch.set_grad_enabled discussion](https://discuss.pytorch.org/t/does-model-eval-with-torch-set-grad-enabled-is-train-have-the-same-effect-for-grad-history/17183)

## 3. `torch.no_grad()`
- Context manager that temporarily disables gradient tracking and prevents the creation of intermediate buffers during inference.
- This leads to improved memory efficiency and faster execution.
- Reference: [PyTorch Performance Tuning Guide](https://docs.pytorch.org/tutorials/recipes/recipes/tuning_guide.html)

**Comment by jackbravo** [comment]
The links are great! Could you add the most relevant to the code? Not everyone checks PRs for important bits of information.

**Comment by Ricardo1232** [comment]
I already added the comments with the links to the PyTorch optimizations, as well as the truncate_dim, which allowed me to remove the get_embedding_dimension() function, simplifying the code.
In this commit: https://github.com/CodeandoGuadalajara/dof-rag/pull/34/commits/3f34568921cece608df352e9946fa8d56466ddd6
